### ISSUE TITLE
Can't use timer tasks on Heroku

### ISSUE THREAD

**Issue by BenjiWilson** [my8100/scrapydweb#139]
It seems that ScrapydWeb can't store timer tasks and other data on a Heroku installation because the code requires creating *databases*, not tables when storing the information. When I try to enter my heroku postgres database URI, I get the message:

psycopg2.OperationalError: FATAL:  database "[username]" does not exist

It appears that ScrapydWeb is attempting to access a database by the same name of the username entered in the Postgres URI as a "DATABASE_URL"

Any thoughts on this? I may try to modify the code so it only creates tables instead of databases.

**Comment by my8100** [comment]
Can you provide the related settings of ScrapydWeb (Use same * to hide sensitive info) and post the logs?

**Comment by BenjiWilson** [comment]
Hi -- thank you for the response. 

I am using the heroku-postgresql add-on. 

In scrapydweb_settings_v10.py, I have tried both:

`DATABASE_URL = os.environ.get('DATABASE_URL', '')`

and 

`DATABASE_URL = 'postgres://username:password@127.0.0.1:5432'`

formats. 

For the latter (manual URI entry), I have tried both with no database specified, and database specified (using postgres://username:password@127.0.0.1:5432/database_name) 

Both result in the error:

```
2020-05-03T01:55:34.339755+00:00 app[web.1]: conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
2020-05-03T01:55:34.339804+00:00 app[web.1]: psycopg2.OperationalError: FATAL:  database "**USERNAME**" does not exist
```

Note that the error states psycopg2 can't find a database with the **username** portion of the URI entry. 

In other words, it appears that ScrapydWeb is looking for a database with the USERNAME i've entered in the postgres URI (and the username yielded from the Heroku DATABASE_URL variable). It's not using the database specified by manual URI definition or from Heroku DATABASE_URL variable. 

Looking at the scrapydweb code, I did see that it attempts to **create** databases, which obviously doesn't work with the heroku postgresql add-on, which auto generates a single database (unless I am missing something on this front).

Thanks again.

**Comment by my8100** [comment]
Below is how I test postgres on CircleCI. 
There is no need to pass the database name in the DATABASE_URL option. 
https://github.com/my8100/scrapydweb/blob/6b9663b275dbb312a97f2819bf657dd0a3cc80c8/.circleci/config.yml#L221-L235

Can you create the databases manually?
You can refer to below code:
https://github.com/my8100/scrapydweb/blob/6b9663b275dbb312a97f2819bf657dd0a3cc80c8/scrapydweb/utils/setup_database.py#L7-L11

https://github.com/my8100/scrapydweb/blob/6b9663b275dbb312a97f2819bf657dd0a3cc80c8/scrapydweb/utils/setup_database.py#L120-L168

**Comment by MartinMetselaar** [comment]
I think this issue still relevant. When running a dyno on Heroku you can add a Heroku Postgres Database. (https://elements.heroku.com/addons/heroku-postgresql. Which is one single database with a preset name. It is not possible to change this name since it looks like this database is running on a server with a lot of other databases.

It is certainly not possible to create multiple databases on 1 server since it is not a database server which we have complete access to.

It however looks like it is possible to add multiple Heroku Postgres add ons but there is no guarantee that they will have the same database url. Each database will get a different username and password pair so there can't be a single DATABASE_URL.

@BenjiWilson Were you able to adjust the code so it creates tables instead of several databases?

**Comment by MartinMetselaar** [comment]
Management to solve it in a different way. Currently running at DigitalOcean a droplet which has a Postgres server installed. You could use https://www.digitalocean.com/community/tutorials/how-to-install-and-use-postgresql-on-ubuntu-18-04 to do the same. And installed nginx as a firewall https://www.digitalocean.com/community/tutorials/how-to-install-nginx-on-ubuntu-18-04 (not sure if it was needed).

But with the above setup it works 👌🏼 since then scrapydweb can create multiple databases at one endpoint.

However a droplet at DigitalOcean is not free.

**Comment by mxdev88** [comment]
Having the same issue. What's the recommended way to use heroku postgres database add-on with scrapydweb?

**Comment by scottatdrake** [comment]
Noting that this is still an issue. 

> [Dynos are also restarted (cycled) at least once per day ](https://devcenter.heroku.com/articles/dynos#restarting)to help maintain the health of applications running on Heroku. Any changes to the local filesystem will be deleted.

When that happens, scrapydweb saved timer tasks are wiped out.

The current Scrapydweb architecture doesn't appear to lend itself well to the way Heroku provides its Postgres addon, which is a single Postgres DB on a shared database server. The scrapydweb app does not have access to create new DBs. (It seems a bit unusual that an app would require four separate databases to persist its data, doesn't it?)

I'm new to Python and scrapy but if I can figure it out, I'll report back. Would love some help if anybody can point me in the right direction.

**Comment by mxdev88** [comment]
I took a stab at it [here](https://github.com/mxdev88/scrapydweb/commit/beec32a03a906f36a703678d165cdaee971c5559). And think one way to fix this issue would be to add an option for single database.  It wouldn't be really complicated to implement though I haven't tested fully.

@my8100 if you think the idea is worth it. I could clean up the code and make it into an actual PR.
